extra/onnx_ops.py

```diff
@@ -263,3 +261,62 @@ def OneHot(indices, depth, values, axis=-1):
 
 def Floor(x:Tensor): return x.floor()
 def Ceil(x:Tensor): return x.ceil()
+
+def EmbedLayerNormalization(input_ids, segment_ids:Optional[Tensor]=None, word_embedding:Tensor=None, position_embedding:Tensor=None, segment_embedding:Optional[Tensor]=None, gamma=None, beta=None, mask:Optional[Tensor]=None, position_ids:Optional[Tensor]=None, epsilon=None, mask_index_type=None):
+  # https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.EmbedLayerNormalization
+  assert (segment_ids is None) is (segment_embedding is None)
+  assert (mask is None) is (mask_index_type is None)
+  assert mask is None, "functionality not supported yet"  # TODO
+  input_shape = input_ids.shape
+  bsz, seq_length = input_shape[0], input_shape[1]
+  compute_seg_emb = (segment_embedding is not None and segment_ids is not None)
+  vocab_size, max_position_embeddings, type_vocab_size = word_embedding.shape[0], position_embedding.shape[0], (segment_embedding.shape[0] if compute_seg_emb else None)
+
+  def embedding(x:Tensor, vocab_size, weight:Tensor)->Tensor:  # TODO from nn.Embedding. Could probably upstream this to Tensor
+    vocab_counter = Tensor.arange(vocab_size, requires_grad=False).reshape(1, 1, vocab_size).expand(*x.shape, vocab_size)
+    return (vocab_counter == x.unsqueeze(2).expand(*x.shape, vocab_size)) @ weight
```

Comment:
If I am not mistaken you want to cast to the dtype of the weight here. Such that weights with float16's are actually computed faster.
```suggestion
    return (vocab_counter == x.unsqueeze(2).expand(*x.shape, vocab_size)).cast(weight.dtype) @ weight
```

Reply:
[953bf04](https://github.com/tinygrad/tinygrad/pull/1537/commits/953bf0497b3aaf650914ba74fe08ac8835660da6)
In theory, ```x.dtype``` is fine also, since input is actually float16. The problem is that currently the weights are loaded as float32 regardless. See ```extra/onnx.py``` line 45. I had a quick look at it, and changing that cast destroyed model accuracy, so this should probably be part of another PR.